smoke.py

```diff
@@ -193,5 +206,73 @@ def testUserManipulation(self):
             self._check_cmd_result(cmd, test_ret, test_output, expect_pass=False)
 
 
+    def _prepare_compiler_test_directory(self):
+
+        # create a temporary directory
+        tmpdir = tempfile.mkdtemp()
+
+        self.log.info("Compiler test temporary directory is %s" % tmpdir)
+
+        # Copy the `hello.sh` script from this resource directory into the
+        # temporary directory
+        src = os.path.join(self.compiler_resource_dir, "hello.sh")
+        dest = os.path.join(tmpdir, "hello.sh")
+        try:
+            shutil.copy(src, dest)
+        except shutil.Error as e:
+            self.log.info('Error: %s' % e)
+        except IOError as e:
+            self.log.info('Error: %s' % e.strerror)
+
+        # make sure destination script is executable
+        st = os.stat(dest)
+        os.chmod(dest, st.st_mode | stat.S_IEXEC)
+
+        # Place a gzipped tarball of `hello.c` and `Makefile` from the
+        # resource directory into the temporary directory with the name
+        # `hello.tgz`.
+        dest = os.path.join(tmpdir, "hello.tgz")
+        tar = tarfile.open(dest, "w:gz")
+        for f in ["hello.c", "Makefile"]:
+            src = os.path.join(self.compiler_resource_dir, f)
+            tar.add(src, arcname=f)
+        tar.close()
+
+        self.compiler_test_dir = tmpdir
+
+    def _cleanup_compiler_test_directory(self):
+        # clean up the temporary directory
+        shutil.rmtree(self.compiler_test_dir)
+
+    def testCompiler(self):
+        """
+        Run a basic C compiler test on our docker image.
+
+        This actually tests the integration of several things, including the
+        ability to install packages, extract a gzipped tarball, run make to
+        compile a very simple C program, and run the compiled executable.
+        """
+
+        self._prepare_compiler_test_directory()
+
+        # mount the prepared compiler test directory in the docker container
+        # as /mnt, then run the "hello.sh" test script within
+        try:
+            test_output = subprocess.check_output(
+                ('docker run -v "%s:/mnt:z" --rm %s bash -c "/mnt/hello.sh"' %
+                 (self.compiler_test_dir, self.br_image_name)),
+                stderr=subprocess.STDOUT, shell=True)
+        except subprocess.CalledProcessError as e:
+            self.error("command '%s' returned exit status %d; output:\n%s" %
+                       (e.cmd, e.returncode, e.output))
+        self.log.info("compiler test succeeded with output:\n%s" % test_output)
```

Comment:
It would likely be prudent to compare the STDOUT content to ensure that we actually get the expected message. (Just in case we discovered an odd compiler bug or locale issue).